### ISSUE TITLE
create a DIFFICULTY LEVEL to be used by the user

### ISSUE THREAD

**Issue by DiNasset22** [shlomif/PySolFC#355]
As you know there are a lot of cards set-ups which do not conduct to a solution.
It could be nice to have different levels of difficulty when starting, like EASY, MEDIUM, DIFFICULT (standard).
For instance, when the user asks for NOT DIFFICULT (EASY or MEDIUM) the algorithm could avoid_
- to place ACES in the last position on a pile
- to place KINGS before ACES on a pile
- etc...

**Comment by joeraz** [comment]
This sounds good on paper, but would be very difficult to implement.

What actually makes a game layout easier or more difficult may vary considerably from game to game.  For example, some games build the foundations down from king, in which case, putting aces in the last position on a pile may actually make the game easier.  And there are other games that start the foundation from an arbitrary rank or has some that build up and others that build down, so where the aces are placed don't really have a clear effect on the game's difficulty.

**Comment by DiNasset22** [comment]
Understand.
In my suggestions I was referring to the game "Quaranta Ladroni" (in
Italian words), which is the one I use.

**Comment by joeraz** [comment]
That would be Forty Thieves, which builds down on the tableau and up from ace on the foundations (will make note of this name for the translation files).  This is probably the most common structure of a builder game, but it's not universal.

**Comment by DiNasset22** [comment]
Yes, that is.
So suggestions as those I made can be valid for a lot of solitaires.
Another general solution (which in this case I believe is used already in
some of the solitaires is allowing the move of sequences, instead of a
single card. That could be of some help in easing the solution.

**Comment by joeraz** [comment]
Yeah, but that would simultaneously have a different effect on a bunch of other games.  And even among games with the "build down on the tableau and up on the foundations" structure, the logic would need to vary a bit based on the different numbers of rows/cards per row.

Moving sequences versus single cards is part of the game rules - making it a variation could mess with statistics and saved games.  We do often do this by having the version where sequences can be moved as a separate game - i.e. "Relaxed FreeCell" for "FreeCell", or "Josephine" for "Forty Thieves".

**Comment by DiNasset22** [comment]
What I was trying to explain it was the idea to find a way of avoiding
frustration in the users:
- *a game cannot be proposed in a way that 99% if not more of them is not
solvable*
Maybe, if you are planning to make use of IA in your development you could
check that -as it is now- that Forty Thieves in not solvable in too many
cases and the user is obliged to skip tenth of schemes before accepting to
start one with the chance to solve in one case out of ten.
Up to you to find a viable solution

**Comment by joeraz** [comment]
One option I've had requested multiple times is an option to only give solvable deals.  The problem is figuring out which games are solvable.  There are two potential ways to do this, but neither would be easy to implement:
- Have a built-in solver for every game, and run every deal through it to determine if the game can be won before presenting it to the player.  A Monte Carlo algorithm could be the best way to do this.
- Collect a list of game deals that we can confirm are solvable, and limit the games the player is given to this list.  Probably easier to code, but we'd need to confirm a few solvable deals for every game (including new ones going forward), and it would result in a much smaller solvable game pool.

If anyone wants to try to add that option, pull requests are always welcome.

But I definitely don't want to go half measures on something like this - like you say, this is about avoiding frustration, and it will do the opposite if the solution only works for some games and not others.

I'll leave this open for further comments or suggestions.

**Comment by hajo4** [comment]
To make it more easy to collect solvable games,
I suggest an option 'autosave-games-on-finish',
and append 'won' or 'lost' to the filename.
Maybe only for 'perfect' games...

Such a list would also be usefull to run demos 
that show how to win a game..
